hibernate-core/src/main/java/org/hibernate/loader/ast/internal/MultiNaturalIdLoaderImpl.java

```diff
@@ -0,0 +1,223 @@
+import org.hibernate.sql.exec.spi.JdbcParameterBindings;
+import org.hibernate.sql.exec.spi.JdbcSelect;
+import org.hibernate.sql.results.LoadingLogger;
+import org.hibernate.sql.results.graph.entity.LoadingEntityEntry;
+import org.hibernate.sql.results.internal.RowTransformerPassThruImpl;
+
+/**
+ * Standard MultiNaturalIdLoader implementation
+ */
+public class MultiNaturalIdLoaderImpl<E> implements MultiNaturalIdLoader<E> {
+
+	// todo (6.0) : much of the execution logic here is borrowed from `org.hibernate.loader.ast.internal.MultiIdEntityLoaderStandardImpl`
+	// 	- consider ways to consolidate/share logic
+
+	private final EntityMappingType entityDescriptor;
+
+	public MultiNaturalIdLoaderImpl(EntityMappingType entityDescriptor, MappingModelCreationProcess creationProcess) {
+		this.entityDescriptor = entityDescriptor;
+	}
+
+	@Override
+	public <K> List<E> multiLoad(K[] naturalIds, MultiNaturalIdLoadOptions options, SharedSessionContractImplementor session) {
+		if ( naturalIds == null ) {
+			throw new IllegalArgumentException( "Natural-ids passed to load is null" );
+		}
+
+		if ( options.isOrderReturnEnabled() ) {
+			return performOrderedMultiLoad( naturalIds, options, session );
+		}
+		else {
+			return performUnorderedMultiLoad( naturalIds, options, session );
+		}
+	}
+
+	private <K> List<E> performOrderedMultiLoad(K[] naturalIds, MultiNaturalIdLoadOptions options, SharedSessionContractImplementor session) {
+		assert options.isOrderReturnEnabled();
+
+		throw new NotYetImplementedFor6Exception( "`#performOrderedMultiLoad` not yet implemented" );
+	}
+
+	private <K> List<E> performUnorderedMultiLoad(K[] naturalIds, MultiNaturalIdLoadOptions options, SharedSessionContractImplementor session) {
+		if ( naturalIds.length == 0 ) {
+			return Collections.emptyList();
+		}
+
+		if ( LoadingLogger.LOGGER.isTraceEnabled() ) {
+			LoadingLogger.LOGGER.tracef( "Starting unordered multi natural-id loading for `%s`", entityDescriptor.getEntityName() );
+		}
+
+		final SessionFactoryImplementor sessionFactory = session.getFactory();
+
+		final List<E> result = CollectionHelper.arrayList( naturalIds.length );
+
+		final LockOptions lockOptions = (options.getLockOptions() == null)
+				? new LockOptions( LockMode.NONE )
+				: options.getLockOptions();
+
+		int numberOfIdsLeft = naturalIds.length;
+
+		final int maxBatchSize;
+		if ( options.getBatchSize() != null && options.getBatchSize() > 0 ) {
+			maxBatchSize = options.getBatchSize();
+		}
+		else {
+			maxBatchSize = session.getJdbcServices().getJdbcEnvironment().getDialect().getDefaultBatchLoadSizingStrategy().determineOptimalBatchLoadSize(
+					entityDescriptor.getNaturalIdMapping().getJdbcTypeCount( sessionFactory.getTypeConfiguration() ),
+					numberOfIdsLeft
+			);
+		}
+
+		int idPosition = 0;
+		while ( numberOfIdsLeft > 0 ) {
+			final int batchSize = Math.min( numberOfIdsLeft, maxBatchSize );
+
+			final Object[] naturalIdsInBatch = new Object[ batchSize ];
```

Comment:
The statement is dependent upon the number of keys to load.  So yeah we could, but that would require quite a few more changes here.  But probably worth it.

Reply:
Can't we reuse the same statement by binding `null` for the last few ids to fill up the batch? Since we use `id IN (?, ?)` which is the same semantically as `id = ? or id = ?` this should work well.

Reply:
It really makes no difference whether we use IN or OR... the "problem" is the same.  I do not like the "padded" solution.  Was actually thinking more what I do with dynamic batching.

Working on this change...